### PR TITLE
ipsec: Fix `XfrmInNoStates` drops on ESK & AKS upgrades

### PR DESCRIPTION
**Pull request by pchaigno**
This pull request fixes IPsec packet drops of type `XfrmInNoStates` that would happen on EKS and AKS for the whole duration of the upgrade (until the last node is upgraded). The root of the issue is that we switched from using CiliumInternalIP to using NodeInternalIP for the IPsec encapsulation and that change isn't properly handled on the upgrade.

To avoid the `XfrmInNoStates` drops, this pull request introduces a 2-steps upgrade process. First, after the usual upgrade, both CiliumInternalIPs and NodeInternalIPs encapsulations will be accepted on ingress, but NodeInternalIPs will still be used for the encapsulation on egress. Then, in a second time, users can switch to `use-cilium-internal-ip-for-ipsec=true` (defaults false) to start using CiliumInternalIPs.

:warning: Note this pull request doesn't fix other drops of type `XfrmOutPolBlock` also happening during the IPsec + EKS/AKS upgrade. Those will be addressed in a separate PR.

### How was this tested?

I performed two types of tests: first with a normal upgrade process and second with a split cluster:
1. The usual upgrade path was tested on 2 clusters of 20 nodes each, with EKS and GKE. The upgrade was performed from Cilium v1.13.0 to this patchset, with the drop-sensitive `migrate-svc` application running during the upgrade. The two-steps upgrade was carried out and, after each step, I checked that no `XfrmInNoStates` drops were reported. The `migrate-svc` application did have drops during the first step because of the aforementioned `XfrmOutPolBlock` drops. It did not have any additional drops afterward. Our connectivity tests suite also passed after each step.
2. I deployed a 3-nodes EKS cluster with a mix of Cilium versions to simulate the impact of a prolonged upgrade. Two nodes ran Cilium v1.13.0, while the third ran this patchset backported on the v1.13 branch. I ran the connectivity tests and confirmed that no drops, neither `XfrmInNoStates` nor `XfrmOutPolBlock`, were reported.

Fixes: https://github.com/cilium/cilium/pull/24773.
Fixes: https://github.com/cilium/cilium/pull/24030.
```release-note
Fix a bug that would cause connectivity drops of type XfrmInNoStates on upgrade when IPsec is enabled with ENI or Azure IPAM mode.
```

### CODE REVIEW

File: pkg/datapath/linux/node.go [hunk 1]
```diff
@@ -962,101 +993,164 @@ func (n *linuxNodeHandler) enableIPsec(newNode *nodeTypes.Node) {
...
+			localNodeInternalIP, err := getV6LinkLocalIP()
+			if err != nil {
+				log.WithError(err).Error("Failed to get local IPv6 for IPsec configuration")
+			}
+
+			for _, cidr := range n.nodeConfig.IPv6PodSubnets {
+				spi, err := ipsec.UpsertIPsecEndpoint(wildcardCIDR, cidr, localIP, wildcardIP, 0, ipsec.IPSecDirIn, zeroMark)
+				upsertIPsecLog(err, "in CiliumInternalIPv6", wildcardCIDR, cidr, spi)
+
+				spi, err = ipsec.UpsertIPsecEndpoint(wildcardCIDR, cidr, localNodeInternalIP, wildcardIP, 0, ipsec.IPSecDirIn, zeroMark)
```
Comment (jschwinger233) [inline pkg/datapath/linux/node.go:1116]: I was wondering if we can skip generating ingress policy for native node IP when `UseCiliumInternalIPForIPsec` is true, but It looks unsafe during upgrade, so be it.

**Review by jschwinger233 — APPROVED** [review]

File: pkg/datapath/linux/node.go (hunk 1 above)
Comment (pchaigno) [reply pkg/datapath/linux/node.go:1116]: Yep, we need this for the upgrade. We can clean it up later, in v1.15, once we're sure everyone has upgraded from the old IPs.

**Review by mhofstetter — APPROVED** [review]
config changes (agent & CLI) look good to me!

**Pull request merged**